quickwit-metastore/src/checkpoint.rs

```diff
@@ -0,0 +1,457 @@
+// Quickwit
+//  Copyright (C) 2021 Quickwit Inc.
+//
+//  Quickwit is offered under the AGPL v3.0 and as commercial software.
+//  For commercial licensing, contact us at hello@quickwit.io.
+//
+//  AGPL:
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Affero General Public License as
+//  published by the Free Software Foundation, either version 3 of the
+//  License, or (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Affero General Public License for more details.
+//
+//  You should have received a copy of the GNU Affero General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use serde::ser::SerializeMap;
+use serde::Deserialize;
+use serde::Serialize;
+use std::cmp::Ordering;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::fmt;
+use std::ops::Range;
+use std::sync::Arc;
+use thiserror::Error;
+use tracing::info;
+use tracing::warn;
+
+/// PartitionId identifies a partition for a given source.
+#[derive(Debug, Default, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct PartitionId(pub Arc<String>);
+
+impl From<String> for PartitionId {
+    fn from(partition_id_str: String) -> Self {
+        PartitionId(Arc::new(partition_id_str))
+    }
+}
+
+impl From<&str> for PartitionId {
+    fn from(partition_id_str: &str) -> Self {
+        PartitionId(Arc::new(partition_id_str.to_string()))
+    }
+}
+
+impl From<u16> for PartitionId {
+    fn from(partition_id: u16) -> Self {
+        let partition_id_str = format!("{:0>5}", partition_id);
+        PartitionId(Arc::new(partition_id_str))
+    }
+}
+
+impl From<u64> for PartitionId {
+    fn from(partition_id: u64) -> Self {
+        let partition_id_str = format!("{:0>20}", partition_id);
+        PartitionId(Arc::new(partition_id_str))
+    }
+}
+
+/// Marks a position within a specific partition of a source.
+///
+/// The nature of the position may very depending on the source.
+/// Each source needs to encode it as a String in such a way that
+/// the lexicographical order matches the natural order of the
+/// position.
+///
+/// For instance, for u64 a 0-left-padded decimal representation
+/// can be used. Alternatively a base64 representation of their
+/// Big Endian representation can be used.
+#[derive(Debug, Default, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct Position(pub Arc<String>);
```

Comment:
In which case do we need `String` for the partition offsets?

Reply:
The source I know of (Kinesis, Kafka, Pulsar) are using either u64, u128.
We use `String` here to stay agnostic.

I guess we could take a risk and use u128, and hope this is ok.